### PR TITLE
🐛 set default value for StandardDestination/SourceDefinition `tombstone` fields

### PR DESCRIPTION
**Pull request by pedroslopez**
## What

`StandardDestinationDefinition` and `StandardSourceDefinition` schemas currently have the `tombstone` field as nullable booleans, however, the database itself has a non-null constraint on the `actor_definitions.tombstone` column. This means that you can't just write a StandardSourceDefinition or StandardDestinationDefinition to the database using something like `ConfigRepository.writeStandardSourceDefinition()` without first doing some patching on the object to ensure that this tombstone field has been set. 

This is the case whenever we try to load definitions from a RemoteDefinitionsProvider into the db. For example:

```
2022-09-13 03:04:15 INFO i.a.b.w.BootloaderAppWrapped(lambda$applyMaskedSeedSpec$2):135 - Applying masked seed specs for Cloud..
2022-09-13 03:04:15 ERROR i.a.b.w.BootloaderAppWrapped(lambda$applyMaskedSeedSpec$2):138 - Wrapped Bootloader load error: 
org.jooq.exception.DataAccessException: SQL [update "public"."actor_definition" set "id" = cast(? as uuid), "name" = ?, "docker_repository" = ?, "docker_image_tag" = ?, "documentation_url" = ?, "icon" = ?, "actor_type" = ?::"public"."actor_type", "source_type" = ?::"public"."source_type", "spec" = cast(? as jsonb), "protocol_version" = ?, "tombstone" = ?, "public" = ?, "custom" = ?, "release_stage" = ?::"public"."release_stage", "release_date" = cast(? as date), "resource_requirements" = cast(? as jsonb), "updated_at" = cast(? as timestamp with time zone) where "public"."actor_definition"."id" = cast(? as uuid)]; ERROR: null value in column "tombstone" of relation "actor_definition" violates not-null constraint
  Detail: Failing row contains (137ece28-5434-455c-8f34-69dc3782f451, LinkedIn Ads, airbyte/source-linkedin-ads, 0.1.9, https://docs.airbyte.io/integrations/sources/linkedin-ads, linkedin.svg, source, api, {"supportsDBT": false, "documentationUrl": "https://docs.airbyte..., 2022-08-01 22:45:16.088923+00, 2022-09-13 03:04:15.410804+00, null, generally_available, null, null, t, f, null).
	at org.jooq_3.13.4.POSTGRES.debug(Unknown Source) ~[?:?]
	at org.jooq.impl.Tools.translate(Tools.java:2753) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:755) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:385) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.AbstractDelegatingQuery.execute(AbstractDelegatingQuery.java:119) ~[jooq-3.13.4.jar:?]
	at io.airbyte.config.persistence.ConfigWriter.lambda$writeStandardSourceDefinition$0(ConfigWriter.java:67) ~[config-persistence-dev-4e7ba06.jar:?]
	at java.util.Collections$SingletonList.forEach(Collections.java:4966) ~[?:?]
	at io.airbyte.config.persistence.ConfigWriter.writeStandardSourceDefinition(ConfigWriter.java:34) ~[config-persistence-dev-4e7ba06.jar:?]
	at io.airbyte.config.persistence.DatabaseConfigPersistence.lambda$writeStandardSourceDefinition$32(DatabaseConfigPersistence.java:853) ~[config-persistence-dev-4e7ba06.jar:?]
	at io.airbyte.db.Database.lambda$transaction$0(Database.java:27) ~[db-lib-dev-4e7ba06.jar:?]
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$0(DefaultDSLContext.java:564) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.Tools$12$1.block(Tools.java:4926) ~[jooq-3.13.4.jar:?]
	at java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:3744) ~[?:?]
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3689) ~[?:?]
	at org.jooq.impl.Tools$12.get(Tools.java:4923) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:616) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:533) ~[jooq-3.13.4.jar:?]
	at io.airbyte.db.Database.transaction(Database.java:27) ~[db-lib-dev-4e7ba06.jar:?]
	at io.airbyte.db.ExceptionWrappingDatabase.transaction(ExceptionWrappingDatabase.java:31) ~[db-lib-dev-4e7ba06.jar:?]
	at io.airbyte.config.persistence.DatabaseConfigPersistence.writeStandardSourceDefinition(DatabaseConfigPersistence.java:852) ~[config-persistence-dev-4e7ba06.jar:?]
	at io.airbyte.config.persistence.DatabaseConfigPersistence.writeConfig(DatabaseConfigPersistence.java:729) ~[config-persistence-dev-4e7ba06.jar:?]
	at io.airbyte.config.persistence.ConfigRepository.writeStandardSourceDefinition(ConfigRepository.java:248) ~[config-persistence-dev-4e7ba06.jar:?]
	at io.airbyte.bootloader.wrapped.ApplySeedToDefinitions.apply(ApplySeedToDefinitions.java:52) ~[io.airbyte-cloud-airbyte-bootloader-wrapped-dev-4e7ba06-170de72.jar:?]
	at io.airbyte.bootloader.wrapped.ApplySeedToDefinitions.apply(ApplySeedToDefinitions.java:42) ~[io.airbyte-cloud-airbyte-bootloader-wrapped-dev-4e7ba06-170de72.jar:?]
	at io.airbyte.bootloader.wrapped.BootloaderAppWrapped.lambda$applyMaskedSeedSpec$2(BootloaderAppWrapped.java:136) ~[io.airbyte-cloud-airbyte-bootloader-wrapped-dev-4e7ba06-170de72.jar:?]
	at io.airbyte.bootloader.BootloaderApp.load(BootloaderApp.java:175) [airbyte-bootloader-dev-4e7ba06.jar:?]
	at io.airbyte.bootloader.wrapped.BootloaderAppWrapped.main(BootloaderAppWrapped.java:108) [io.airbyte-cloud-airbyte-bootloader-wrapped-dev-4e7ba06-170de72.jar:?]
Caused by: org.postgresql.util.PSQLException: ERROR: null value in column "tombstone" of relation "actor_definition" violates not-null constraint
  Detail: Failing row contains (137ece28-5434-455c-8f34-69dc3782f451, LinkedIn Ads, airbyte/source-linkedin-ads, 0.1.9, https://docs.airbyte.io/integrations/sources/linkedin-ads, linkedin.svg, source, api, {"supportsDBT": false, "documentationUrl": "https://docs.airbyte..., 2022-08-01 22:45:16.088923+00, 2022-09-13 03:04:15.410804+00, null, generally_available, null, null, t, f, null).
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675) ~[postgresql-42.3.5.jar:42.3.5]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365) ~[postgresql-42.3.5.jar:42.3.5]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355) ~[postgresql-42.3.5.jar:42.3.5]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490) ~[postgresql-42.3.5.jar:42.3.5]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408) ~[postgresql-42.3.5.jar:42.3.5]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:167) ~[postgresql-42.3.5.jar:42.3.5]
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:156) ~[postgresql-42.3.5.jar:42.3.5]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44) ~[HikariCP-5.0.1.jar:?]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java) ~[HikariCP-5.0.1.jar:?]
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:209) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:453) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.AbstractDMLQuery.execute(AbstractDMLQuery.java:904) ~[jooq-3.13.4.jar:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:371) ~[jooq-3.13.4.jar:?]
	... 23 more
```

(from setting the env vars in https://github.com/airbytehq/airbyte-cloud/pull/2732)

## How

We could do some patching to set the tombstone field, but I think the right thing to do here is just set a proper default value on the StandardSource/DestinationDefinition schemas since this behavior is unexpected. Other boolean fields, like `public` and `custom` have defaults set on them so they don't have this issue.

So, my solution here is:
- Set `default: false` on StandardDestinationDefinition and StandardSourceDefinition schemas

### CODE REVIEW

File: airbyte-config/init/src/main/java/io/airbyte/config/init/RemoteDefinitionsProvider.java [hunk 1]
```diff
@@ -50,10 +50,10 @@ public void initialize() throws InterruptedException, IOException {
     final CombinedConnectorCatalog catalog = getRemoteDefinitionCatalog(this.remoteDefinitionCatalogUrl, this.timeout);
     this.sourceDefinitions = catalog.getSources().stream().collect(Collectors.toMap(
         StandardSourceDefinition::getSourceDefinitionId,
-        source -> source));
+        source -> source.withTombstone(source.getTombstone() != null && source.getTombstone())));
```
Comment (alafanechere) [inline airbyte-config/init/src/main/java/io/airbyte/config/init/RemoteDefinitionsProvider.java:53]: I find this logic difficult to understand that it falls back to `false` when tombstone is `null`. 

nit: What do you think about keeping a bunch of definition static patching methods in a helper class + defining a `patchDefinition` method in this class that will call these helper functions? 
I would separate concerns and keep definition patching outside of initialize: 

```java
    this.sourceDefinitions = catalog.getSources().stream().collect(Collectors.toMap(
        StandardSourceDefinition::getSourceDefinitionId,
        source -> patchDefinition(source)));
    this.destinationDefinitions = catalog.getDestinations().stream().collect(Collectors.toMap(
        StandardDestinationDefinition::getDestinationDefinitionId,
        source -> patchDefinition(source)));
```

Comment (pedroslopez) [reply airbyte-config/init/src/main/java/io/airbyte/config/init/RemoteDefinitionsProvider.java]: 👍🏻 moved to the existing JsonDefinitionsHelper class

File: airbyte-config/init/src/main/java/io/airbyte/config/init/JsonDefinitionsHelper.java [hunk 2]
```diff
@@ -7,9 +7,28 @@
...
 
 public class JsonDefinitionsHelper {
 
+  public static StandardSourceDefinition patchSourceDefinition(final StandardSourceDefinition sourceDefinition) {
+    final JsonNode jsonNode = Jsons.jsonNode(sourceDefinition);
+    addMissingTombstoneField(jsonNode);
+    addMissingPublicField(jsonNode);
+    addMissingCustomField(jsonNode);
+    return Jsons.object(jsonNode, StandardSourceDefinition.class);
+  }
```
Comment (alafanechere) [inline airbyte-config/init/src/main/java/io/airbyte/config/init/JsonDefinitionsHelper.java:22]: The definition models have setters ( eg: `setTombstone`, I find it awkward to require an intermediate cast to JsonNode to add these fields.

Comment (pedroslopez) [reply airbyte-config/init/src/main/java/io/airbyte/config/init/JsonDefinitionsHelper.java]: moving to just setting a default on the standardsourcedefinition and standarddestinationdefinition as this is unexpected behavior to me and all this stuff shouldn't be necessary imo

**Comment by pedroslopez** [comment]
Hey @cgardens, do you have context on why the tombstone field _didn't_ have a default? Is there a reason for this? Tests seem to pass fine after adding this default.

**Comment by pedroslopez** [comment]
Oh, noticed @pmossman actually added this in https://github.com/airbytehq/airbyte/pull/8823 so tagging him instead

**Review by pmossman — APPROVED** [review]
Looks good! i agree it makes sense to set an explicit default value to false for tombstone fields.

**Pull request merged**